Add integration for igloohome devices

keithle888 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This adds an integration for users of [igloohome](https://www.igloohome.co/) devices that regularly polls the last known battery level of devices that they own.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/35784

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. https://github.com/keithle888/igloohome-api/releases/tag/v0.0.5

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

keithle888 committed:
- Initial creation of integration

keithle888 committed:
- Added corrected auth API

keithle888 committed:
- Created working flow that uses client credentials but has at error at the end

keithle888 committed:
- Added lock platform

keithle888 committed:
- Switched out to sensor platform to monitor battery levels.
- Switched out to using igloohome-api library

keithle888 committed:
- Fixed entity ID

keithle888 committed:
- Added fault tolerance on platform setup

keithle888 committed:
- Exceptions are re-raised during home assistant standards

keithle888 committed:
- Added a scan interval of 1 hour

keithle888 committed:
- updated scan interval for platform

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,67 @@
+        raise PlatformNotReady from e
+
+
+class BatteryBasedDevice(BaseEntity, SensorEntity):
```
```suggestion
class IglooHomeBatterySensor(BaseEntity, SensorEntity):
```

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,67 @@
+class BatteryBasedDevice(BaseEntity, SensorEntity):
+    """Implementation of a device that has a battery."""
+
+    _attr_native_unit_of_measurement = "%"
+
+    def __init__(self, device_id: str, device_name: str, type: str, api: Api) -> None:
+        """Initialize the class."""
+        super().__init__(
+            device_id=device_id, device_name=device_name, type=type, api=api
+        )
+        self._attr_device_class = SensorDeviceClass.BATTERY
```
```suggestion
    _attr_native_unit_of_measurement = "%"
    _attr_device_class = SensorDeviceClass.BATTERY

    def __init__(self, device_id: str, device_name: str, type: str, api: Api) -> None:
        """Initialize the class."""
        super().__init__(
            device_id=device_id, device_name=device_name, type=type, api=api
        )
```

zweckj commented on homeassistant/components/igloohome/entity.py:
```diff
@@ -0,0 +1,31 @@
+
+
+class BaseEntity(Entity):
+    """A base entity that is a part of all igloohome devices."""
+
```
```suggestion
    """A base entity that is a part of all igloohome devices."""
    
   _attr_has_entity_name = True

```
must be set for new integrations

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,67 @@
+        """Update the battery level."""
+        try:
+            response = await self.api.get_device_info(deviceId=self.device_id)
+            self._attr_native_value = response.batteryLevel
+        except Exception as e:
+            raise HomeAssistantError from e
```
```suggestion
        except Exception as e:
            raise HomeAssistantError from e
        else:
             self._attr_native_value = response.batteryLevel
```

zweckj commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,21 @@
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
+    }
```
not used

zweckj commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,21 @@
+    "step": {
+      "user": {
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "username": "[%key:common::config_flow::data::username%]",
+          "password": "[%key:common::config_flow::data::password%]"
```
actually you only have `client_id` and `client_secret`

keithle888 committed:
- Renamed MyConfigEntry to IgloohomeConfigEntry

keithle888 committed:
- Removed comments

keithle888 committed:
- Removed unnecessary early returns

keithle888 committed:
- Moved code that won't fail out of try block

keithle888 committed:
Update homeassistant/components/igloohome/sensor.py

Moved SensorDeviceClass.Battery out of __init__

Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

keithle888 committed:
Apply suggestions from code review

commit suggestions from PR

Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

keithle888 committed:
- Renamed BaseEntity to IgloohomeBaseEntity

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
Update homeassistant/components/igloohome/entity.py

Added has name attribute.

Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
- Minor fixes from apply suggestions

keithle888 committed:
- Removed empty keys from manifest.json

keithle888 committed:
- Renamed BatteryBasedDevice to IgloohomeBatteryEntity

keithle888 committed:
- Removed unused strings

keithle888 committed:
Merge branch 'dev' into feature/igloohome

keithle888 committed:
- Changed dependency for getting client session from hassio to helper module

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
- Stripped out generated config_flow test. Created for initially generated API code that has been replaced.

keithle888 committed:
- bump igloohome-api dependency to 0.0.5

keithle888 committed:
Merge branch 'dev' into feature/igloohome

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
- Re-included test for the config flow form.

keithle888 committed:
- Added test for authentication failure

keithle888 committed:
- Added test for connection failure

keithle888 committed:
- Updated BatteryEntity to only be populated when a batteryLevel is provided in /devices API.

zweckj commented on homeassistant/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,36 @@
+async def async_setup_entry(hass: HomeAssistant, entry: IgloohomeConfigEntry) -> bool:
+    """Set up igloohome from a config entry."""
+
+    session = async_get_clientsession(hass)
+    authentication = Auth(
+        session=session,
```
```suggestion
    authentication = Auth(
        session=async_get_clientsession(hass),
```
no need for that var

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,92 @@
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    session = aiohttp.ClientSession()
+    auth = Auth(
+        session=session,
```
```suggestion
    auth = Auth(
        session=aiohttp.ClientSession(),
```

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    return {"title": "Client Credentials"}
+
+
+class MyConfigFlow(ConfigFlow, domain=DOMAIN):
```
still relevant

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
```
what I mean is, if you put the validate method here in its entirety, you can do without the extra exceptions

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,72 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Do setup the lock entities."""
+
+    try:
+        api: Api = entry.runtime_data
```
```suggestion
    api= entry.runtime_data
    try:
```

zweckj commented on homeassistant/components/igloohome/entity.py:
```diff
@@ -0,0 +1,33 @@
+
+    _attr_has_entity_name = True
+
+    def __init__(self, device_id: str, device_name: str, type: str, api: Api) -> None:
+        """Initialize the base device class."""
+        self.device_id = device_id
+        self.device_name = device_name
```
just pass and store the entire device here, makes things a bit less complicated

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,72 @@
+    entry: IgloohomeConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Do setup the lock entities."""
```
```suggestion
    """Set up sensor entities."""
```

zweckj commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,14 @@
+  "config": {
+    "step": {
+      "user": {
+        "data": {}
```
you'll need to fill that

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,139 @@
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with patch(
```
better to use a fixture for that

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,139 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_auth(
```
those two tests can be parametrized

zweckj reviewed (changes requested):
please don't request a review unless you either fixed or addressed all the open points

keithle888 committed:
- Removed unnecessary variable declarations

keithle888 committed:
- Moved unnecessary statement out of try block

keithle888 committed:
- Set device info as attribute in __init__.

keithle888 committed:
- Merge validation and handling user input in config_flow code to reduce custom error classes

keithle888 committed:
- Introduced fix where errors is not yet assigned

keithle888 committed:
- Added strings to user config flow to provide more clarity to the step

keithle888 committed:
- re-wrote invalid form input tests to use input parameterization

keithle888 commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+
+@parameterized.expand(
+    [(AuthException(), "invalid_auth"), (ClientError(), "cannot_connect")]
```
Having issues trying to run parameterized test. No test collected.

zweckj replied:
try `@pytest.mark.parametrize()` instead

zweckj commented on homeassistant/components/igloohome/entity.py:
```diff
@@ -0,0 +1,31 @@
+    @property
+    def device_info(self) -> dr.DeviceInfo:
+        """Return the device info."""
+        return dr.DeviceInfo(
```
yes I mean in init. There is no real difference, the properties return the `_attr_...` by default. We just like the attrs for everything that's static.

keithle888 committed:
- Fixed parameterized test

keithle888 committed:
- Added successful auth as test fixture

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,56 @@
+                client_secret=user_input[CONF_CLIENT_SECRET],
+            )
+            try:
+                _ = await auth.async_get_access_token()
```
```suggestion
                await auth.async_get_access_token()
```

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,72 @@
+) -> None:
+    """Set up sensor entities."""
+
+    api: Api = entry.runtime_data
```
```suggestion
    api = entry.runtime_data
```
type is inferred

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,72 @@
+        raise PlatformNotReady from e
+    else:
+        async_add_entities(
+            new_entities=(
+                IgloohomeBatteryEntity(
+                    device_id=device.deviceId,
+                    device_name=device.deviceName,
+                    type=device.type,
+                    api=api,
+                )
+                for device in devices_response.payload
+                if device.batteryLevel is not None
+            ),
```
```suggestion
                IgloohomeBatteryEntity(
                    device_id=device.deviceId,
                    device_name=device.deviceName,
                    type=device.type,
                    api=api,
                )
                for device in devices_response.payload
                if device.batteryLevel is not None
```

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,72 @@
+        """Update the battery level."""
+        try:
+            response = await self.api.get_device_info(deviceId=self.device_id)
+        except Exception as e:
```
we don't allow catching bare Exceptions outside of the config flow

zweckj commented on homeassistant/components/igloohome/entity.py:
```diff
@@ -0,0 +1,30 @@
+
+    _attr_has_entity_name = True
+
+    def __init__(self, device_id: str, device_name: str, type: str, api: Api) -> None:
```
I'd rather pass the entire device. It's more future proof and save on unnecessary parameters.

keithle888 replied:
Hey @zweckj , subclasses needs at least the `device_id` to query it's status with the API. Pulling this out of the device seems counter productive as they would need to know how the `DeviceInfo.identifier` is constructed.

Also, including more platforms in the future which would then create subclasses of this entity. Therefore, different entities of the same physical device will easily have matching DeviceInfo objects constructed.

zweckj replied:
yes, but why not pass the entire device, but all those different pieces?  if you pass the entire devices it's cleaner and future platforms have access to any properties of that device

zweckj replied:
I'm not talking about the `DeviceInfo`. The devices your passing is here is your device type from `devices_response = await api.get_devices()`, the `DeviceInfo` is then built in this constructor here, which is fine.

What I don't like is that you are picking properties out of _your_ devices. This would mean if you'd e.g. like to add a serial number to the `DeviceInfo` later, you'd need to add yet another parameter to this function. Instead if you pass and store your device to the entities, you'll be less limited later.

```suggestion
    def __init__(self, device: <your_device_type>, api: Api) -> None:
```

keithle888 replied:
Hey @zweckj , thanks for the clarification and can make the changes.

I had imagined that this current structure helps decouple the dependency of the `IgloohomeEntity` from the API's `Device`. 

Should the API change in the future, the changes necessary would then purely fall to just updating the mappings from the new API data structure to the entities constructor, and would not need to change all the internal workings of the entity that had they relied the API's data structure.

zweckj replied:
I'd still think it'd be better to pass the entire thing. If it changes you'll have to update it in `entity.py` and the platforms attributes and not all the constructors all over the place.

zweckj reviewed (changes requested):
- since last week you'll also need to add a `quality_scale.yaml` and need to at least reach "Bronze" from those criteria (feel free to ask if you have any questions) https://developers.home-assistant.io/docs/core/integration-quality-scale/
- I'd also highly suggest to add snapshot tests for your sensors, can be easily done with the helper `await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)`

keithle888 commented:
> * quality_scale.yaml

Hello @zweckj , where can I find the `quality_scale.yaml` file and do you mean it needs to be added to the component?

keithle888 committed:
- Minor correction

keithle888 committed:
- Minor syntax changes from github suggestions

keithle888 committed:
- Moved away from catching Exception

keithle888 committed:
Merge branch 'dev' into feature/igloohome

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
Merge branch 'home-assistant:dev' into feature/igloohome

keithle888 committed:
Created quality_scale.yaml

zweckj commented on homeassistant/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,35 @@
+        client_secret=entry.data[CONF_CLIENT_SECRET],
+    )
+
+    entry.runtime_data = Api(auth=authentication)
```
```suggestion
    entry.runtime_data = IgloohomeApi(auth=authentication)
```
would make it a bit more descriptive imo

zweckj commented on homeassistant/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,35 @@
+
+from __future__ import annotations
+
+from igloohome_api import Api, Auth
```
```suggestion
from igloohome_api import Api, Auth as IglooHomeApi, IglooHomeAuth
```

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,73 @@
+    api = entry.runtime_data
+
+    try:
+        devices_response = await api.get_devices()
```
it probably makes sense to do that step in __init__ `async_setup_entry` and store the result in `entry.runtime_data` so you don't have to get the devices for every platform. This will also help you satisfy ` test-before-setup`

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,73 @@
+            response = await self.api.get_device_info(deviceId=self.device_id)
+        except (ApiException, ClientError) as e:
+            raise HomeAssistantError from e
+        else:
+            self._attr_native_value = response.batteryLevel
```
```suggestion
        self._attr_native_value = response.batteryLevel
```
we'll not reach that otherwise anyways

zweckj commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,19 @@
+        "data": {
+          "client_id": "Client ID",
+          "client_secret": "Client Secret"
+        }
```
```suggestion
        },
        "data_description": {
             "client_id": "<Some description what this is>",
              "client_secret": "<Some description what this is>",
         }
```
please also add data descriptions

zweckj commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,19 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Client Credentials",
```
Make this a bit more descriptive

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  has-entity-name: done
+  runtime-data: done
+  test-before-configure: done
+  test-before-setup: done
```
is currently todo

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  runtime-data: done
+  test-before-configure: done
+  test-before-setup: done
+  unique-config-entry: todo
```
you'll need to satisfy all the requirements in "Bronze", so fix the things in bronze that are still "todo"

keithle888 committed:
- Updated constructor of entity to use data class from the API
- Updated API/Auth names to IgloohomeApi/IgloohomeAuth for clarity
- Added pulling of devices in the __init__ module instead of platform

keithle888 committed:
- Minor syntax cleanup

keithle888 committed:
- Added data_description for Config Entry fields

keithle888 committed:
Updated title to be more descriptive

keithle888 committed:
Updated exception handling

keithle888 committed:
- Set unique id to config flow

keithle888 committed:
- Added abort on duplicate config flow id

keithle888 committed:
- Updated config flow descriptions

keithle888 committed:
- marked some bronze reqs as done

keithle888 committed:
- marked some bronze reqs as done

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on homeassistant/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,51 @@
+        raise ConfigEntryAuthFailed from e
+    except (ApiException, ClientError) as e:
+        raise ConfigEntryNotReady from e
+    else:
+        entry.runtime_data = (api, devices)
```
```suggestion
    entry.runtime_data = (api, devices)
```
you don't reach that otherwise

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+)
+
+
+class MyConfigFlow(ConfigFlow, domain=DOMAIN):
```
```suggestion
class IglooHomeConfigFlow(ConfigFlow, domain=DOMAIN):
```

zweckj replied:
You didn’t solve that one

keithle888 replied:
Apologies, forgot to push the changes. :3

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+                errors["base"] = "cannot_connect"
+            else:
+                await self.async_set_unique_id(
+                    f"{user_input[CONF_CLIENT_ID]}_{user_input[CONF_CLIENT_SECRET]}"
```
with those two you could check earlier. But I don't think that's a good unique_id. This would probably allow setting the same hub (account?) multiple times. Is there something you can get from the account you can use for uniqueness?

keithle888 replied:
Hey @zweckj , unfortunately, the `userinfo` endpoint for OAuth2 is also not available with the client credentials. So this is the best option I can find at the moment.

zweckj replied:
are the devices connected to a Hub you can maybe get a serial, etc. from?

keithle888 replied:
@zweckj , I've switched over to rejecting the user input based on matching user input. 

To respond to your question, I unfortunately can't see serial numbers being an effective provider of uniqueness. The config entry is tied to a user's account, therefore the number of devices can change and so can the number of hubs which only effectively acts as a bluetooth bridge.

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  config-entry-unloading: done
+  docs-configuration-parameters: todo
+  docs-installation-parameters: todo
+  entity-unavailable: done
```
```suggestion
  entity-unavailable: todo
```

keithle888 replied:
Updated the sensor to set availability.

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  docs-supported-functions: todo
+  docs-troubleshooting: todo
+  docs-use-cases: todo
+  dynamic-devices:
```
but you are polling for multiple devices in init and setting up `sensor for device`? So what actually is it?

keithle888 replied:
Apologies, missed that and marked it as `todo`

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+  stale-devices: todo
+  # Platinum
+  async-dependency: done
+  inject-websession: done
```
```suggestion
  inject-websession: todo
```

keithle888 replied:
I'm currently using injecting the `ClientSession`s from `async_get_clientsession()` which fulfills this rule, or am I not understanding it correctly?

zweckj replied:
I somehow missed that, disregard

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,68 @@
+        """Initialize the class."""
+        super().__init__(api_device_info=api_device_info, api=api)
+        # Set the unique ID of the battery entity.
+        self._attr_unique_id = f"battery_{api_device_info.deviceId}"
```
I would rather pass a `key` to your baseentity (`battery` in this case) and construct the unique id there

keithle888 replied:
Done, used `SensorDeviceClass.BATTERY` as the `key`, is that okay? 

On a separate note, what is the normal practice here for resolving conversations? Should the PR author be doing it here or should that be left to the reviewer?

zweckj replied:
You can resolve that if you believe it’s addressed, no worries

zweckj reviewed (changes requested):
I'd still suggest you add a snapshot test for your sensor platform and your device. Take a look here, how quickly this can be done https://github.com/home-assistant/core/blob/dev/tests/components/acaia/test_sensor.py#L20-L31

keithle888 committed:
- Set sensor to update availability if API fails

keithle888 committed:
- Moved setting of entity unique id to base class

keithle888 committed:
- Updated some rules as todo

keithle888 committed:
- Minor rename of config flow

keithle888 committed:
- Minor refactor

keithle888 commented:
> I'd still suggest you add a snapshot test for your sensor platform and your device. Take a look here, how quickly this can be done https://github.com/home-assistant/core/blob/dev/tests/components/acaia/test_sensor.py#L20-L31

Hey @zweckj , thanks for the link. Snapshot testing is quite new to me plus the structure of the tests/methods used is quite foreign. This will take awhile before I get a grasp of it and implement something here.

keithle888 commented:
> I'd still suggest you add a snapshot test for your sensor platform and your device. Take a look here, how quickly this can be done https://github.com/home-assistant/core/blob/dev/tests/components/acaia/test_sensor.py#L20-L31

Hey @zweckj , would you be able to provide a hint on how to patch async functions in a class?

zweckj commented:
> > I'd still suggest you add a snapshot test for your sensor platform and your device. Take a look here, how quickly this can be done https://github.com/home-assistant/core/blob/dev/tests/components/acaia/test_sensor.py#L20-L31
> 
> Hey @zweckj , would you be able to provide a hint on how to patch async functions in a class?

Either you `MagicMock` the entire class with `autospec=True` or you assign  `my_func= AsyncMock()`

keithle888 committed:
- Added snapshot test for a single lock sensor

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on homeassistant/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,50 @@
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+type IgloohomeConfigEntry = ConfigEntry[(IgloohomeApi, GetDevicesResponse)]
```
I'd rather prefer a `dataclass` for the content of the ConfigEntry (instead of a tuple)

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,72 @@
+    )
+
+
+class IgloohomeBatteryEntity(IgloohomeBaseEntity, SensorEntity):
```
are you really going to add only Battery as sensor (in the foreseeable future)?

keithle888 replied:
@zweckj , yes. Batteries are the only sensor entity available from the company's API and was the simpliest integration available that is a useful feature. The company's main products are battery-powered locks and so I plan to add the Lock entity down the road. However, that feature is more complicated and did not seem like the best option for a first integration.

zweckj commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,84 @@
+            return_value=GetDevicesResponse(
+                nextCursor="",
+                payload=[
+                    GetDeviceInfoResponse(
+                        id="123456",
+                        type="Lock",
+                        deviceId="OE1X123cbb11",
+                        deviceName="Front Door",
+                        pairedAt="2024-11-09T11:19:25+00:00",
+                        homeId=[],
+                        linkedDevices=[],
+                        batteryLevel=100,
```
you are already reusing that, put in a const

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with auth_successful:
```
```suggestion
```
should be unnecessary

keithle888 replied:
Hey @zweckj . By removing this, fixture does not patch the Auth class and the test ends up trying to execute a HTTP request that fails. Not quite sure why that happens though... Any ideas? 

Attached the failing test output.

```
============================== 1 failed in 0.42s ===============================
FAILED                                    [100%]INFO:homeassistant.loader:Loaded igloohome from homeassistant.components.igloohome

tests/components/igloohome/test_config_flow.py:15 (test_form)
hass = <HomeAssistant RUNNING>
mock_setup_entry = <AsyncMock name='async_setup_entry' id='281472760325440'>
auth_successful = <unittest.mock._patch object at 0xffff7b909850>

    async def test_form(
        hass: HomeAssistant, mock_setup_entry: AsyncMock, auth_successful
    ) -> None:
        """Test we get the form."""
        result = await hass.config_entries.flow.async_init(
            DOMAIN, context={"source": config_entries.SOURCE_USER}
        )
        assert result["type"] == FlowResultType.FORM
        assert result["errors"] == {}
    
>       result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_CLIENT_ID: "test-client-id",
                CONF_CLIENT_SECRET: "test-client-secret",
            },
        )

test_config_flow.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../homeassistant/data_entry_flow.py:341: in async_configure
    result = await self._async_configure(flow_id, user_input)
../../../homeassistant/data_entry_flow.py:388: in _async_configure
    result = await self._async_handle_step(
../conftest.py:761: in _flow_manager_async_handle_step
    result = await _original_flow_manager_async_handle_step(self, flow, *args)
../../../homeassistant/data_entry_flow.py:491: in _async_handle_step
    result: _FlowResultT = await getattr(flow, method)(user_input)
../../../homeassistant/components/igloohome/config_flow.py:44: in async_step_user
    await auth.async_get_access_token()
../../../venv/lib/python3.12/site-packages/igloohome_api/__init__.py:63: in async_get_access_token
    response = await self.session.post(
../../../venv/lib/python3.12/site-packages/aiohttp/client.py:701: in _request
    conn = await self._connector.connect(
../../../venv/lib/python3.12/site-packages/aiohttp/connector.py:544: in connect
    proto = await self._create_connection(req, traces, timeout)
../../../venv/lib/python3.12/site-packages/aiohttp/connector.py:1050: in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
../../../venv/lib/python3.12/site-packages/aiohttp/connector.py:1363: in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
../../../venv/lib/python3.12/site-packages/aiohttp/connector.py:1109: in _wrap_create_connection
    sock = await aiohappyeyeballs.start_connection(
../../../venv/lib/python3.12/site-packages/aiohappyeyeballs/impl.py:109: in start_connection
    raise first_exception
../../../venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:127: in run_one_coro
    result = await coro_fn()
../../../venv/lib/python3.12/site-packages/aiohappyeyeballs/impl.py:151: in _connect_sock
    sock = socket.socket(family=family, type=type_, proto=proto)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pytest_socket.disable_socket.<locals>.GuardedSocket'>
family = <AddressFamily.AF_INET: 2>, type = <SocketKind.SOCK_STREAM: 1>
proto = 6, fileno = None

    def __new__(cls, family=-1, type=-1, proto=-1, fileno=None):
        if _is_unix_socket(family) and allow_unix_socket:
            return super().__new__(cls, family, type, proto, fileno)
    
>       raise SocketBlockedError()
E       pytest_socket.SocketBlockedError: A test tried to use socket.socket.

../../../venv/lib/python3.12/site-packages/pytest_socket.py:89: SocketBlockedError
WARNING:tests.conftest:Lingering timer after test <TimerHandle when=137813.398080583 _set_result(<Future finished result=None created at /usr/local/lib/python3.12/asyncio/base_events.py:449>) at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:20 created at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:152>
WARNING:tests.conftest:Lingering timer after test <TimerHandle when=137813.398995083 _set_result(<Future finished result=None created at /usr/local/lib/python3.12/asyncio/base_events.py:449>) at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:20 created at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:152>
WARNING:tests.conftest:Lingering timer after test <TimerHandle when=137813.399968458 _set_result(<Future finished result=None created at /usr/local/lib/python3.12/asyncio/base_events.py:449>) at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:20 created at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:152>
WARNING:tests.conftest:Lingering timer after test <TimerHandle when=137813.401272958 _set_result(<Future finished result=None created at /usr/local/lib/python3.12/asyncio/base_events.py:449>) at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:20 created at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:152>
WARNING:tests.conftest:Lingering timer after test <TimerHandle when=137813.402235791 _set_result(<Future finished result=None created at /usr/local/lib/python3.12/asyncio/base_events.py:449>) at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:20 created at /IdeaProjects/homeassistant-core/venv/lib/python3.12/site-packages/aiohappyeyeballs/_staggered.py:152>

Process finished with exit code 1
```

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+    with auth_successful:
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_CLIENT_ID: "test-client-id",
+                CONF_CLIENT_SECRET: "test-client-secret",
+            },
```
reused, put in var

keithle888 committed:
- Converted ConfigEntry to data class

keithle888 committed:
- Set constant for repeated use of GetDeviceInfoResponse in tests

keithle888 committed:
- Replaced repeated form input with constant

keithle888 committed:
- Switched out using unique_id for config entry to matching entries instead.

keithle888 commented:
Hey @zweckj , 

Couldn't reply on your thread about the unique_id of the config flow. A user unfortunately can have no hub or multiple hubs so it won't work via that method. I've also read up about the unique ID and figured out that it isn't a good way to go because it has to remain stable. I've swapped out the check to use `_async_abort_entries_match` instead.

keithle888 committed:
Merge branch 'dev' into feature/igloohome

keithle888 committed:
- Added abort text for matching config entry

zweckj commented on homeassistant/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    try:
+        devices = (await api.get_devices()).payload
+    except AuthException as e:
+        raise ConfigEntryAuthFailed from e
```
```suggestion
        raise ConfigEntryError from e
```
we don't have a reauth flow atm, so raise this until it's introduced (in a follow up PR)

zweckj commented on homeassistant/components/igloohome/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+            self._async_abort_entries_match(
+                {
+                    CONF_CLIENT_ID: user_input[CONF_CLIENT_ID],
+                    CONF_CLIENT_SECRET: user_input[CONF_CLIENT_SECRET],
```
```suggestion
```
even the client id should be sufficient here

zweckj commented on homeassistant/components/igloohome/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/igloohome",
+  "iot_class": "cloud_polling",
+  "requirements": ["igloohome-api==0.0.5"]
```
we require proper releases on your repo (with changelog and tag), so please add one there

keithle888 replied:
Done, tags are already present. Added a release. I've also added a github action that creates releases automatically for future updates.

zweckj commented on homeassistant/components/igloohome/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/igloohome",
+  "iot_class": "cloud_polling",
+  "requirements": ["igloohome-api==0.0.5"]
```
```suggestion
  "quality_scale": "bronze",
  "requirements": ["igloohome-api==0.0.5"]
```

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+    comment: |
+      This integration does not provide additional actions.
+  docs-high-level-description: done
+  docs-installation-instructions: done
```
currently you're missing a {% include integrations/config_flow.md %}
and also a {% configuration_basic %} block describing the config parameters

keithle888 replied:
@zweckj , are you referring to the documentation for `homeassistant.io`?

keithle888 replied:
Done.

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  unique-config-entry: done
+
+  # Silver
+  action-exceptions: todo
```
currently exempt, no actions and read only platform

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+
+  # Silver
+  action-exceptions: todo
+  config-entry-unloading: todo
```
```suggestion
  config-entry-unloading: done
```

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,68 @@
+  exception-translations: todo
+  icon-translations: todo
+  reconfiguration-flow: todo
+  repair-issues: todo
```
exempt, no issues atm

zweckj commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,26 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "API credentials is already in use."
```
```suggestion
      "already_configured": "Client ID is already in use."
```

zweckj commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,73 @@
+        yield mock_setup_entry
+
+
+@pytest.fixture
+def auth_successful():
+    """Set up the Auth module to always successfully operate."""
+    return patch(
+        "igloohome_api.Auth.async_get_access_token",
+        return_value="mock_access_token",
+    )
```
```suggestion
@pytest.fixture
async def mock_auth() -> Generator[MagicMock]:
    """Mock the auth class."""
    with (
            patch(
            "igloohome_api.Auth.",
            autospec=True
        ) as mock_auth,
        patch(
            "igloohome_api.config_flow.Auth",
            new=mock_auth,
        )
    ):
        auth = mock_auth.return_value
        auth.async_get_access_token.return_value = True
        yield auth
```
I'd mock away the entire object. This should also solve your problem below you were asking about

keithle888 replied:
Hey @zweckj , I've tried integrating your suggested changes here but ran into an issue that I'm trying to figure out. Do you what the issue is here?

(Needed to switch the patch targets to the `IgloohomeAuth` objects being constructed in the `igloohome` component modules)

```
FAILED                                    [100%]INFO:homeassistant.loader:Loaded igloohome from homeassistant.components.igloohome

tests/components/igloohome/test_config_flow.py:20 (test_form)
hass = <HomeAssistant RUNNING>
mock_setup_entry = <AsyncMock name='async_setup_entry' id='281473424716768'>
mock_auth = <NonCallableMagicMock name='IgloohomeAuth()' spec='Auth' id='281473422491520'>

    async def test_form(
        hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth
    ) -> None:
        """Test we get the form."""
        result = await hass.config_entries.flow.async_init(
            DOMAIN, context={"source": config_entries.SOURCE_USER}
        )
        assert result["type"] == FlowResultType.FORM
        assert result["errors"] == {}
    
>       with mock_auth:
E       TypeError: 'NonCallableMagicMock' object does not support the context manager protocol

test_config_flow.py:31: TypeError

Process finished with exit code 1
```

zweckj replied:
you don't use that with `with`, but just add it to the parameters of your test

keithle888 replied:
Thanks for the hint.

zweckj commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,73 @@
+        domain=DOMAIN,
+        version=1,
+        data={"client_id": "client_id", "client_secret": "client_secret"},
+        unique_id="aa:bb:cc:dd:ee:ff",
```
```suggestion
```
doesn't have a unique id

zweckj commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,73 @@
+
+
+@pytest.fixture(autouse=True)
+def api_single_lock():
```
missing return type and I'd just call that `mock_api` as you might want to modify the response in later tests, so the name wouldn't be true anymore

keithle888 replied:
Hey @zweckj , I'd like to not keep it generic as my intent is to keep the fixture's name descriptive to the "state" the mock is representing. 

I feel that a generic `mock_api` is not descriptive of what can be expected of the mock, so I would like to keep the name and re-use this in the future to re-present more possible states.

zweckj replied:
Since this is the one which is autoused I suggested `mock_api` because that is going to be the default state for all your tests. If you then need to manipulate you can just do `mock_api.property = something` and the name is still true

keithle888 replied:
Ah okay, with `autouse` that would make alot of sense. Thanks.

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,84 @@
+
+
+@pytest.mark.parametrize(
+    ("auth_exception", "result_error"),
```
```suggestion
    ("exception", "result_error"),
```

keithle888 committed:
Replaced exception based on suggestion.

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Added quality_scale in manifest

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Adapted already_configured string

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Update homeassistant/components/igloohome/quality_scale.yaml

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
- Updated action-setup quality_scale as exempt

keithle888 committed:
- Updated repair-issues quality_scale as exempt

keithle888 committed:
Removed client_secret from config entry abort check

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
Removed redundant unique_id.

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
- Minor cleanup

keithle888 committed:
Replaced auth_succesful mock with mock_auth

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
- Rolled back changes to auth_successful fixture

keithle888 committed:
- Removed api_single_lock to api_mock

keithle888 committed:
- Renamed back to api_mock_single_lock

keithle888 committed:
- Minor rename

keithle888 committed:
- Minor rename of the mock_api since it is autouse

keithle888 committed:
Merge branch 'dev' into feature/igloohome

keithle888 commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,70 @@
+def mock_api() -> Generator[AsyncMock]:
+    """Set up the Api module. Defaults to always returning a single lock."""
+    with (
+        patch(
+            "homeassistant.components.igloohome.IgloohomeApi",
+            autospec=True,
+        ) as api_mock,
```
Hey @zweckj , do you know why I can't patch `igloohome_api.Api` here similar to  the `mock_auth()` fixture? My `test_sensor.py` test fails with an error that a real API call is being make (socket.connect()).

zweckj replied:
I don't know without seeing the code.

keithle888 replied:
Hmm, okay. I'll try to figure it out separately.

zweckj reviewed (changes requested):
your library has a dependency conflict and does not install

keithle888 committed:
- Bump igloohome-api version to 0.0.6 to solve dependency conflicts.

keithle888 commented:
> your library has a dependency conflict and does not install

Thanks. Sorted it out :)

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,84 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
```
why do you not use the firxture here

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,84 @@
+    # Make sure the config flow tests finish with either an
+    # FlowResultType.CREATE_ENTRY or FlowResultType.ABORT so
+    # we can show the config flow is able to recover from an error.
+    with patch(
```
or here

keithle888 committed:
- Replaced with mocks in tests with external fixtures

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,78 @@
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with mock_auth:
```
since the fixture appears to be working in the test below you don't need that here, params is enough

zweckj commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,71 @@
+
+
+@pytest.fixture
+async def mock_auth():
```
```suggestion
async def mock_auth() -> Generator[AsyncMock]:
```

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,78 @@
+
+
+async def test_form(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth
```
```suggestion
    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth: Generator[AsyncMock]
```

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,78 @@
+    [(AuthException(), "invalid_auth"), (ClientError(), "cannot_connect")],
+)
+async def test_form_invalid_input(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth, exception, result_error
```
```suggestion
    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth: Generator[AsyncMock], exception: Exception, result_error: str
```

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,78 @@
+    # Make sure the config flow tests finish with either an
+    # FlowResultType.CREATE_ENTRY or FlowResultType.ABORT so
+    # we can show the config flow is able to recover from an error.
+    mock_auth.side_effect = "mock_access_token"
```
```suggestion
    mock_auth.side_effect = None
```

zweckj commented on tests/components/igloohome/test_config_flow.py:
I'm missing a test for checking the abort is working correctly

keithle888 replied:
Hey @zweckj , what is abort in the context of a config flow?

keithle888 replied:
Ah okay, you're talking about aborting when two config entries match.

keithle888 replied:
Hey @zweckj , 

running into this strange missing string error even though the string is already present. Got any tips off the top of your head?

```
/usr/local/lib/python3.12/asyncio/base_events.py:687: in run_until_complete
    return future.result()
../../../venv/lib/python3.12/site-packages/pytest_asyncio/plugin.py:333: in async_finalizer
    await gen_obj.__anext__()  # type: ignore[union-attr]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

ignore_translations = []

    @pytest.fixture(autouse=True)
    async def check_translations(
        ignore_translations: str | list[str],
    ) -> AsyncGenerator[None]:
        """Check that translation requirements are met.
    
        Current checks:
        - data entry flow results (ConfigFlow/OptionsFlow/RepairFlow)
        - issue registry entries
        """
        if not isinstance(ignore_translations, list):
            ignore_translations = [ignore_translations]
    
        translation_errors = {k: "unused" for k in ignore_translations}
    
        translation_coros = set()
    
        # Keep reference to original functions
        _original_flow_manager_async_handle_step = FlowManager._async_handle_step
        _original_issue_registry_async_create_issue = ir.IssueRegistry.async_get_or_create
        _original_service_registry_async_call = ServiceRegistry.async_call
    
        # Prepare override functions
        async def _flow_manager_async_handle_step(
            self: FlowManager, flow: FlowHandler, *args
        ) -> FlowResult:
            result = await _original_flow_manager_async_handle_step(self, flow, *args)
            await _check_config_flow_result_translations(
                self, flow, result, translation_errors
            )
            return result
    
        def _issue_registry_async_create_issue(
            self: ir.IssueRegistry, domain: str, issue_id: str, *args, **kwargs
        ) -> None:
            result = _original_issue_registry_async_create_issue(
                self, domain, issue_id, *args, **kwargs
            )
            translation_coros.add(
                _check_create_issue_translations(self, result, translation_errors)
            )
            return result
    
        async def _service_registry_async_call(
            self: ServiceRegistry,
            domain: str,
            service: str,
            service_data: dict[str, Any] | None = None,
            blocking: bool = False,
            context: Context | None = None,
            target: dict[str, Any] | None = None,
            return_response: bool = False,
        ) -> ServiceResponse:
            try:
                return await _original_service_registry_async_call(
                    self,
                    domain,
                    service,
                    service_data,
                    blocking,
                    context,
                    target,
                    return_response,
                )
            except HomeAssistantError as err:
                translation_coros.add(
                    _check_exception_translation(self._hass, err, translation_errors)
                )
                raise
    
        # Use override functions
        with (
            patch(
                "homeassistant.data_entry_flow.FlowManager._async_handle_step",
                _flow_manager_async_handle_step,
            ),
            patch(
                "homeassistant.helpers.issue_registry.IssueRegistry.async_get_or_create",
                _issue_registry_async_create_issue,
            ),
            patch(
                "homeassistant.core.ServiceRegistry.async_call",
                _service_registry_async_call,
            ),
        ):
            yield
    
        await asyncio.gather(*translation_coros)
    
        # Run final checks
        unused_ignore = [k for k, v in translation_errors.items() if v == "unused"]
        if unused_ignore:
            pytest.fail(
                f"Unused ignore translations: {', '.join(unused_ignore)}. "
                "Please remove them from the ignore_translations fixture."
            )
        for description in translation_errors.values():
            if description not in {"used", "unused"}:
>               pytest.fail(description)
E               Failed: Translation not found for igloohome: `config.abort.already_configured`. Please add to homeassistant/components/igloohome/strings.json

../conftest.py:916: Failed
```

keithle888 replied:
Forgot to run `python3 -m script.translations develop --all`.

keithle888 replied:
Nope, I had pushed some changes to the test that failed. It's working now after updating the translations.

zweckj commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,69 @@
+        super().__init__(
+            api_device_info=api_device_info,
+            api=api,
+            unique_key=SensorDeviceClass.BATTERY,
```
```suggestion
            unique_key="battery",
```
styling/personal preference, but I'd rather have an actual string here

keithle888 committed:
- Minor improvements

keithle888 committed:
- Minor improvement

keithle888 committed:
- Set unique key to hardcoded value

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
- Added test for attempting config flow with credentials twice

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on homeassistant/components/igloohome/quality_scale.yaml:
```diff
@@ -0,0 +1,74 @@
+    comment: |
+      Integration has no actions and is a read-only platform.
+  docs-configuration-parameters: todo
+  docs-installation-parameters: todo
```
```suggestion
  docs-installation-parameters: done
```

keithle888 committed:
Updated docs-installation-parameters as done

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,94 @@
+    assert len(mock_setup_entry.mock_calls) == 1
+
+    # Attempt another config flow with the same client credentials
+    # and ensure that FlowResultType.ABORT is returned.
```
Ah you added it here. Please make it into a separate test.

keithle888 replied:
Done! Merry Christmas :)

keithle888 committed:
- Split test of abort flow into a separate test with recovery checks.

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,131 @@
+    await hass.async_block_till_done()
+    assert result["type"] == FlowResultType.ABORT
+
+    # Ensure we can recover from an ABORT
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    new_input = FORM_USER_INPUT.copy()
+    new_input[CONF_CLIENT_ID] = "some-other-client-id"
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        new_input,
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Client Credentials"
+    assert result["data"] == new_input
+    assert len(mock_setup_entry.mock_calls) == 2
```
`ABORT` is terminal, so you can't recover here
```suggestion
```

zweckj commented on tests/components/igloohome/__init__.py:
```diff
@@ -0,0 +1,14 @@
+async def setup_integration(
+    hass: HomeAssistant, mock_config_entry: MockConfigEntry
+) -> None:
+    """Set up the acaia integration for testing."""
```
```suggestion
    """Set up the igloohome integration for testing."""
```

zweckj commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,131 @@
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth: Generator[AsyncMock]
+) -> None:
+    """Tests where we handle errors in the config flow."""
+    # Create first config flow.
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        FORM_USER_INPUT,
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Client Credentials"
+    assert result["data"] == FORM_USER_INPUT
+    assert len(mock_setup_entry.mock_calls) == 1
```
```suggestion
setup_integration(hass, mock_config_entry)
```
you'll need `mock_config_entry` in the params and might need to remove `mock_setup_entry` from them (but please try with them first)

keithle888 replied:
Hey @zweckj , wasn't able to get it to work. The test failed because the entry was created despite the matching the user input and the fixture.

zweckj commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,71 @@
+        title="Client Credentials",
+        domain=DOMAIN,
+        version=1,
+        data={"client_id": "client_id", "client_secret": "client_secret"},
```
```suggestion
        data={CONF_CLIENT_ID: "client_id", CONF_CLIENT_SECRET: "client_secret"},
```

keithle888 committed:
- Removed recovery check from ABORT

keithle888 committed:
Merge remote-tracking branch 'origin/feature/igloohome' into feature/igloohome

keithle888 committed:
- Minor improvement

keithle888 committed:
Merge branch 'dev' into feature/igloohome

keithle888 committed:
Minor change

Co-authored-by: Josef Zweck <josef@zweck.dev>

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj reviewed (approved):
LGTM 👍 Let's see if someone else has other remarks.

joostlek commented on homeassistant/components/igloohome/sensor.py:
```diff
@@ -0,0 +1,69 @@
+            )
+        except (ApiException, ClientError) as e:
+            self._attr_available = False
+            raise HomeAssistantError from e
```
I don't believe we re-raise an exception when we can't update

keithle888 replied:
Removed error being used on async_update(). Silently fails and sets entity as unavailable.

joostlek commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,26 @@
+        "description": "Set your account's API keys to enable Home Assistant to poll the status of your devices.",
+        "data": {
+          "client_id": "Client ID",
+          "client_secret": "Client Secret"
```
```suggestion
          "client_secret": "Client secret"
```

joostlek commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,26 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Add API Access Credentials",
+        "description": "Set your account's API keys to enable Home Assistant to poll the status of your devices.",
```
I'd remove the title and set the description to something more helpful like where we can find those API keys for example

joostlek commented on homeassistant/components/igloohome/strings.json:
```diff
@@ -0,0 +1,26 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "Client ID is already in use."
```
I'd go with the default reference, which shows "this account is already configured", which is IMO more clear about what happened than "is already in use". e.g. Did someone else use it?

joostlek commented on tests/components/igloohome/conftest.py:
```diff
@@ -0,0 +1,72 @@
+async def mock_auth() -> Generator[AsyncMock]:
+    """Set up the mock for the igloohome_api Auth module. Defaults to always successfully operate."""
+    with patch(
+        "igloohome_api.Auth.async_get_access_token",
```
Please patch this at the places where we use this, so `homeassistant.components.igloohome.....`

keithle888 replied:
Sure thing. Curious question: Is this a common practice for python/HA? As it would seem cleaner (code-wise) to patch the library directly so that all future usages are also automatically patched rather than targeting the individual modules.

joostlek replied:
https://docs.python.org/3/library/unittest.mock.html#where-to-patch

We also just consider it a good practice to patch where we use it, this way we reduce the risk of side effects. There are integrations which share a lib for example

keithle888 replied:
Thanks a bunch for the link and clarification!

joostlek commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,106 @@
+}
+
+
+async def test_form(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth: Generator[AsyncMock]
+) -> None:
+    """Test we get the form."""
```
Stale test name and docstring

joostlek commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,106 @@
+
+
+async def test_form(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth: Generator[AsyncMock]
```
```suggestion
    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_auth: AsyncMock
```

joostlek commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,106 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        FORM_USER_INPUT,
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.FORM
    assert result["errors"] == {}

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        FORM_USER_INPUT,
    )

    assert result["type"] is FlowResultType.CREATE_ENTRY
```

joostlek replied:
Please also change this in the other tests

joostlek commented on tests/components/igloohome/test_config_flow.py:
```diff
@@ -0,0 +1,106 @@
+        FORM_USER_INPUT,
+    )
+    await hass.async_block_till_done()
+    assert result["type"] == FlowResultType.ABORT
```
Let's also assert the reason

joostlek commented:
Also, will you add more platforms in the future? It looks like a good fit for the lock entity.

keithle888 committed:
- Removed error being used on async_update(). Silently fails but sets entity as unavailable.

keithle888 committed:
- Minor change

keithle888 committed:
Updated description for user config flow to be more descriptive of the action being done.

keithle888 committed:
- Replaced already_configured error message with default string

keithle888 committed:
- Updated to patch usage targets rather than the library?

keithle888 committed:
- Updated test name

keithle888 committed:
Corrected tests according to suggested changes.

keithle888 committed:
- Added test for reason in matching_entry test

keithle888 commented:
> Also, will you add more platforms in the future? It looks like a good fit for the lock entity.

Yup, that's will be my next objective. I chose an easier-to-implement functionlity as a stepping stone into HA development.

keithle888 committed:
Merge branch 'dev' into feature/igloohome

zweckj reviewed (changes requested):
> > Also, will you add more platforms in the future? It looks like a good fit for the lock entity.
> 
> Yup, that's will be my next objective. I chose an easier-to-implement functionlity as a stepping stone into HA development.

I think what joostlek was aiming for: Since you're going to implement more platforns, but your endpoint look device- not feature-based it probably make sense to implement a `DataUpdateCoordinator` for centralized updates (intead of in the platforms)

keithle888 commented:
> > > Also, will you add more platforms in the future? It looks like a good fit for the lock entity.
> > 
> > 
> > Yup, that's will be my next objective. I chose an easier-to-implement functionlity as a stepping stone into HA development.
> 
> I think what joostlek was aiming for: Since you're going to implement more platforns, but your endpoint look device- not feature-based it probably make sense to implement a `DataUpdateCoordinator` for centralized updates (intead of in the platforms)

Ah okay, thanks for the clarification. The current APIs for (un)locking does not work the same way as a battery level where we can poll for the lock's state. The API requires that I trigger jobs to be executed by a bridge, and either listen for callbacks or poll for the job's state.

I'm going to have to play around with either a `Lock` entity or with [actions](https://developers.home-assistant.io/docs/integration_events) to see what fits best with the API.

keithle888 committed:
Merge branch 'dev' into feature/igloohome

joostlek commented:
Do you have discord? Can you send you discord username?

keithle888 commented:
> Do you have discord? Can you send you discord username?

Hey @joostlek , yep. It would be `keithle`
